### ISSUE TITLE
Overzealous -Wshorten-int-to-cap / loss of integer provenance?

### ISSUE THREAD

**Issue by nwf** [CTSRD-CHERI/llvm-project#588]
`snmalloc` does some bit-packing within a `uintptr_t` (that needs to keep pointer provenance).  This basically Just Works (TM), a testament to the efforts taken for CHERI C compatibility.  However, in one case, it looks like `clang` gets a little confused; manually minimizing a bit,,,
```
struct MetaEntry {
  uintptr_t remote_and_sizeclass;

  sizeclass_t get_sizeclass() const
  {
    return remote_and_sizeclass & (alignof(RemoteAllocator) - 1);
  }
}
```

Here, `clang` reports:
```
/cheri/source/mainline/snmalloc2/src/backend/../mem/../mem/metaslab.h:286:35: error: implicit conversion loses capability metadata: 'unsigned __intcap' to 'snmalloc_enclave::sizeclass_t' (aka 'unsigned long') [-Werror,-Wshorten-cap-to-int]
      return remote_and_sizeclass & (alignof(RemoteAllocator) - 1);
```
which suggests that either `-Wshorten-int-to-cap` is being overzealous or that `clang` has failed to conclude (or has forgotten) that the result of the binary `&` is definitely an integer and not a capability, despite being typed as a `uintptr_t`.  I had originally converted that to be `static_cast<sizeclass_t>(remote_and_sizeclass) & ...`, which works (that is, it is both correct and does not elicit a `-Wshorten-int-to-cap` warning), but Alex points out that I probably should not have had to do that.

**Comment by jrtc27** [comment]
Yeah in this particular case it happens to be mostly clear, but where do you draw the line? You could have a huge alignof, or a tiny address at run time, and then might actually get a valid capability out of the masking... so I think a hard line is the right thing to do here, otherwise it becomes inconsistent and people will keep wanting the analysis to get smarter and smarter as they hit "obvious" cases of false positives.

**Comment by jrtc27** [comment]
Agreed for that example, if we've proved a definite lack of provenance then we can skip the warning.
